sopel/tools/jobs.py

```diff
@@ -127,75 +199,284 @@ def _call(self, job):
+    precedence, and the :meth:`execute` method won't interfere::
+
+        with job:
+            # the job is now running, you can perform pre-execute action
+            job.execute()  # execute the job's action, no state modification
+            # the job is still marked as "running"
+            # you can perform post-execute action
 
-    max_catchup = 5
-    """How many seconds the job can get behind.
+        # outside of the with statement, the job is not running anymore
+
+    .. seealso::
+
+        The :class:`sopel.plugins.jobs.Scheduler` class is specifically
+        designed for plugin's jobs, expecting an instance of
+        :class:`sopel.bot.Sopel` as a manager, and should be used to manipulate
+        plugin jobs.
+
+        In all other case, the :class:`sopel.tools.jobs.Scheduler` class is a
+        generic job scheduler.
 
-    This governs how much the scheduling of jobs is allowed to get behind
-    before they are simply thrown out to avoid calling the same function too
-    many times at once.
     """
+    @classmethod
+    def kwargs_from_callable(cls, handler):
+        """Generate the keyword arguments to create a new instance.
+
+        :param callable handler: callable used to generate keyword arguments
+        :return: a map of keyword arguments
+        :rtype: dict
+
+        This classmethod takes the ``handler``'s attributes to generate a map
+        of keyword arguments for the class. This can be used by the
+        :meth:`from_callable` classmethod to instantiate a new rule object.
 
-    def __init__(self, interval, func):
-        self.next_time = time.time() + interval
-        self.interval = interval
-        self.func = func
+        The expected attributes are the ones set by decorators from the
+        :mod:`sopel.module` module.
+        """
+
+        return {
+            'plugin': getattr(handler, 'plugin_name', None),
+            'label': getattr(handler, 'rule_label', None),
+            'threaded': getattr(handler, 'thread', True),
+            'doc': inspect.getdoc(handler),
+        }
+
+    @classmethod
+    def from_callable(cls, settings, handler):
+        """Instantiate a Job from the bot's ``settings`` and a ``handler``.
+
+        :param settings: bot's settings
+        :type settings: :class:`sopel.config.Config`
+        :param handler: callable used to instantiate a new job
+        :type handler: :term:`function`
+        """
+        kwargs = cls.kwargs_from_callable(handler)
+        return cls(
+            set(handler.interval),
+            handler=handler,
+            **kwargs)
+
+    def __init__(self,
+                 intervals,
+                 plugin=None,
+                 label=None,
+                 handler=None,
+                 threaded=True,
+                 doc=None):
+        # scheduling
+        now = time.time()
+        self.intervals = set(intervals)
+        """Set of intervals to execute the job at."""
```

Comment:
```suggestion
        """Set of intervals at which to execute the job."""
```

"Don't end a sentence with a preposition," they say.